src/gluonnlp/adapters.py

```diff
@@ -0,0 +1,256 @@
+import math
+from collections import OrderedDict
+import mxnet as mx
+from mxnet import np, npx
+from mxnet import use_np
+from mxnet.gluon import nn, HybridBlock, Parameter, Constant
+import numpy as _np
+from .attention_cell import MultiHeadAttentionCell
+import mxnet.numpy_extension as _mx_npx
+from .layers import get_activation, get_norm_layer, _gen_repr_with_kwargs
+
+
+__all__ = ['BasicAdapter', 'AdapterFusion', 'AdapterModule', 'PositionwiseFFN_adapter']
+
+@use_np
+class BasicAdapter(nn.HybridBlock):
+    def __init__(self, config: dict, in_units: int):
+        super().__init__()
+        self._units = config['units']
+        self._activation = config['activation']
+        self.down_proj = nn.Dense(in_units=in_units,
+                                  units=self._units,
+                                  flatten=False,
+                                  weight_initializer=None, bias_initializer='zero')
+        self.activate = get_activation(self._activation)
+        self.up_proj = nn.Dense(in_units=self._units,
+                                  units=in_units,
+                                  flatten=False,
+                                  weight_initializer=None, bias_initializer='zero')
+
+
+
+    def forward(self, data, residual):
+        out = self.down_proj(data)
+        out = self.activate(out)
+        out = self.up_proj(out)
+        return out + residual
+
+
+@use_np
+class AdapterFusion(nn.HybridBlock):
+    def __init__(self, in_units):
+        super().__init__()
+        self.query_proj = nn.Dense(in_units=in_units, units=in_units, flatten=False, weight_initializer=None, bias_initializer='zero')
+        self.key_proj = nn.Dense(in_units=in_units, units=in_units, flatten=False, weight_initializer=None, bias_initializer='zero')
+        self.value_proj = nn.Dense(in_units=in_units, units=in_units, flatten=False, weight_initializer=None, bias_initializer='zero')
+        '''
+        self.attention_cell = MultiHeadAttentionCell(query_units=in_units,
+                                                     num_heads=1,
+                                                     attention_dropout=0,
+                                                     scaled=True)
+        '''
+
+    def forward(self, query, key, value):
+        #query bs, length, unit
+        #key bs, length, num_adapters, unit
+

```

Comment:
Hi xingjian @sxjscience ,could you please check these lines with the utilization of einsum?  I show the original implementation in comment. And the purpose of these lines are similiar to https://github.com/Adapter-Hub/adapter-transformers/blob/0fe1c19f601b7785273e173d30a9392e407823d1/src/transformers/adapters/modeling.py#L211 from line 211 to line 223

Reply:
It looks good to me. One improvement is that there is no need to transpose anymore. You can rely on einsum to fuse these operations in a single op.

Reply:
Ok. Thanks!